overlord/ifacestate/apparmorprompting/common/patterns.go

```diff
@@ -0,0 +1,478 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package common
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+
+	doublestar "github.com/bmatcuk/doublestar/v4"
+)
+
+var ErrNoPatterns = errors.New("no patterns given, cannot establish precedence")
+
+var (
+	// The following matches valid path patterns. Patterns must begin with '\'
+	// and cannot contain unescaped '[' or ']' characters.
+	allowablePathPatternRegexp = regexp.MustCompile(`^/([^\[\]]|\\[\[\]])*$`)
+
+	// The default previously-expanded prefixes to which new patterns or
+	// expanded groups are concatenated. This must be a slice containing the
+	// empty string, since at the beginning of the pattern, we have only one
+	// prefix to which to concatenate, and that prefix is the empty string.
+	// Importantly, this cannot be an empty slice, since concatenating every
+	// entry in an empty slice with every entry in a slice of expanded patterns
+	// would again result in an empty slice.
+	defaultPrefixes = []string{""}
+)
+
+// ExpandPathPattern expands all groups in the given path pattern.
+//
+// Groups are enclosed by '{' '}'. Returns a list of all the expanded path
+// patterns, or an error if the given pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	if len(pattern) == 0 {
+		return nil, fmt.Errorf(`invalid path pattern: pattern has length 0`)
+	}
+	if strings.HasSuffix(pattern, `\`) && !strings.HasSuffix(pattern, `\\`) {
+		return nil, fmt.Errorf(`invalid path pattern: trailing unescaped '\' character: %q`, pattern)
+	}
+	reader := strings.NewReader(pattern)
+	currPrefixes := defaultPrefixes
+	currLiteralStart := 0
+	for {
+		r, _, err := reader.ReadRune()
+		if err != nil {
+			// No more runes
+			break
+		}
+		if r == '\\' {
+			// Skip next rune.
+			reader.ReadRune() // Since suffix is not '\\', must have next rune
+			continue
+		}
+		if r == '}' {
+			return nil, fmt.Errorf(`invalid path pattern: unmatched '}' character: %q`, pattern)
+		}
+		if r != '{' {
+			continue
+		}
+		// Saw start of new group, so get the string from currLiteralStart to
+		// the opening '{' of the new group. Do this before expanding.
+		infix := prevStringFromIndex(reader, currLiteralStart)
```

Comment:
just wondering, why not the standard FSM based recursive parser with a queue of patterns or expansion?

Reply:
I considered using a queue of subpatterns, but it seemed more difficult to reason about which subpatterns to concatenate later. I believe how this works with a reader reading through the pattern string once is essentially a FSM, but I didn't see any other way to look back and extract a substring up to a given point. So either look forward to next `'{', ',', '}'` and add those runes to a string builder, or look back once one is seen and use a slice over the existing string. I think these are basically equivalent, and I wanted to minimize memory copying as much as possible, so I decided to look back and extract the slice.

I also considered an approach to construct a list of lists, where each list contains all the subpatterns in a group, and the set of expanded patterns would be the Cartesian product of the list of lists. But then you have to recursively resolve groups within groups in order to flatten the lists of subpatterns. Basically, I opted for the current approach because I believed the cognitive load would be minimal, but I agree there are some messy parts as a result of trying to maintain good performance.

I think storing a recursive list of the start and end indices of subgroups, and then building each final string using a string Builder might decrease the asymptotic runtime and memory usage by a factor of len(pattern) if creating substring slices involved copying the bytes, but afaik it only creates a header with a start pointer and length, so it's equivalent to storing start and end indices but much easier (IMO) to reason about.

The inefficiency with my current approach is that the previous substrings are concatenated into new strings, and that involves copying. So a more efficient approach would not concatenate any strings until the very end, then doing so with a builder for every permutation of subpatterns. Would this be what you mean by using a queue of patterns or expansion?

I'm very open to any preferences or suggestions you might have. Thanks!